### PR TITLE
Final graph interface accepting adjacency list and incidence matrix

### PR DESCRIPTION
**Pull request by sarrah-basta**
This PR includes the interface for the user to enter any graph that can be represented in the following ways :
enum graph_type {
  GRAPH_ADJ_LIST_UNDIRECTED_UNWEIGHTED,
  GRAPH_ADJ_LIST_UNDIRECTED_WEIGHTED,
  GRAPH_ADJ_LIST_DIRECTED_UNWEIGHTED,
  GRAPH_ADJ_LIST_DIRECTED_WEIGHTED,
  GRAPH_INC_MATRIX_UNDIRECTED_UNWEIGHTED,
  GRAPH_INC_MATRIX_UNDIRECTED_WEIGHTED,
  GRAPH_INC_MATRIX_DIRECTED_UNWEIGHTED,
  GRAPH_INC_MATRIX_DIRECTED_WEIGHTED,
};
and then converted to adjacency matrix in the backend, post-which it is converted to a 1d memref.

### CODE REVIEW

**Review by meshtag — CHANGES_REQUESTED** [review]
General guidelines : 
1. Please format all relevant files with `clang-format`. 
2. Write `doxygen` style documentation for all functions explaining each parameter's use and function description. Keep in mind that we plan to use `doxygen` for automated docs generation. 

Ping me or @amanchhaparia once you are done with above tasks.

**Comment by sarrah-basta** [comment]
@meshtag all required changes are made and committed, please review the PR.

File: include/Interface/Container.h [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+//===----------------------------------------------------------------------===//
+
+
+#ifndef INTERFACE_GRAPH_CORE_CONTAINER_H
+#define INTERFACE_GRAPH_CORE_CONTAINER_H
+
+#include <memory>
+#include <stdint.h>
+#include <vector>
+#include <list>
```
Comment (amanchhaparia) [inline include/Interface/Container.h]: What's the need of `list` header file, did I missed the use-case of it somewhere?

File: lib/Interface/GraphContainer.cpp [hunk 2]
```diff
@@ -0,0 +1,536 @@
...
+/**
+ * @brief Converts the various Graph representations types to a linear memref descriptor
+ * 
+ * @tparam T represents the datatype used
+ * @tparam N represents the number of dimensions.
+ * @return MemRef_descriptor returns a memref descriptor of the respective graph representation type after conversion.
+ */
+
+template <typename T, size_t N> 
+MemRef_descriptor Graph<T, N>::graph_to_MemRef_descriptor()
```
Comment (amanchhaparia) [inline lib/Interface/GraphContainer.cpp]: Didn't understood the logic behind using the `MemRef_descriptor` when you have already created your custom `container.h`, and also creating the graph class by inheriting it.

File: lib/Interface/GraphContainer.cpp [hunk 3]
```diff
@@ -0,0 +1,536 @@
...
+			std::cout<<"Unknown graph type"<<std::endl;
+			break;
+	}
+
+	this->aligned = linear;
+
+	MemRef_descriptor sample_graph_memref = MemRef_Descriptor(this->allocated, this->aligned, this->offset,
+							this->sizes, this->strides);
+
+	return sample_graph_memref;
```
Comment (amanchhaparia) [inline lib/Interface/GraphContainer.cpp]: Kindly use the proper naming conventions.

**Review by amanchhaparia — CHANGES_REQUESTED** [review]
- Remove the empty `CMakeLists.txt` from the **./include/Interface/CMakeLists.txt**.
- @Ris-Bali, add the instructions to generate and view the docs on the local system in the `Readme.md` file.

**Review by meshtag — CHANGES_REQUESTED** [review]
Add all explicit documentation related files (generated after using `doxygen`) in `.gitignore`

File: include/Interface/Container.h (hunk 1 above)
Comment (sarrah-basta) [reply include/Interface/Container.h]: Removed in latest [commit](https://github.com/meshtag/GraphMLIR/pull/19/commits/5ca64fe5dc76d07f58460c41c64c019a5baae26c) accordingly.

File: lib/Interface/GraphContainer.cpp (hunk 2 above)
Comment (sarrah-basta) [reply lib/Interface/GraphContainer.cpp]: Resolved and loaded all required data into an object of the MemRef class in this [commit](https://github.com/meshtag/GraphMLIR/pull/19/commits/5ca64fe5dc76d07f58460c41c64c019a5baae26c). Please have a look @amanchhaparia @meshtag .

**Comment by Ris-Bali** [comment]
> Add all explicit documentation related files (generated after using `doxygen`) in `.gitignore`

Have added the doc folder in .gitignore

**Comment by Ris-Bali** [comment]
> * Remove the empty `CMakeLists.txt` from the **./include/Interface/CMakeLists.txt**.
> * @Ris-Bali, add the instructions to generate and view the docs on the local system in the `Readme.md` file.

Have added the instructions to the readme.

**Comment by amanchhaparia** [comment]
@sarrah-basta, @Ris-Bali, I had build your implementation and tried to pass it in our example `graph_bfs` method, I got the below error(Segmentation fault) kindly solve it ASAP! Also do not mark the PR as ready for review if the testing is not done.

image

**Comment by sarrah-basta** [comment]
> @sarrah-basta, @Ris-Bali, I had build your implementation and tried to pass it in our example `graph_bfs` method, I got the below error(Segmentation fault) 

Resolved in [this commit](https://github.com/meshtag/GraphMLIR/pull/19/commits/f5b4f6e0dcc13ac1ccf85acf603a59d594236620). Please have a look @meshtag @amanchhaparia

**Pull request closed without merging**